File: 300. Longest Increasing Subsequence/Solution300.java
import java.util.Arrays;

class Solution300 {
  public static void main(String[] args) {
    Solution sol = new Solution();
    //int[] nums = {4, 10, 4, 3, 8, 9};
    int[] nums = {0, 1, 0, 3, 2, 3};
    System.out.println(sol.lengthOfLIS(nums));
  }
}

class Solution {
  public int lengthOfLIS(int[] nums) {
    int[] dp = new int[nums.length];
    Arrays.fill(dp, 1);
    int res = 1;
    for (int start = nums.length - 1; start >= 0; start--) {
      for (int end = start + 1; end < nums.length; end++) {
        if (nums[end] <= nums[start]) continue;
        dp[start] = Math.max(dp[start], 1 + dp[end]);
        res = Math.max(res, dp[start]);
      }
    }
    return res;
  }
}

File: 300. Longest Increasing Subsequence/go/solution300.go
package main

import "fmt"

func main() {
	nums := []int{
		10,
		9,
		2,
		5,
		3,
		4,
		101,
		18,
	}

	fmt.Println(lengthOfLIS(nums))
}

func lengthOfLIS(nums []int) int {
	n := len(nums)

	cache := make([][]int, n)
	for i := range cache {
		cache[i] = make([]int, n)
		for j := range cache[i] {
			cache[i][j] = -1
		}
	}

	var dfs func(curr, last int) int

	dfs = func(curr, last int) int {
		if curr >= n {
			return 0
		}
		if cache[curr][last+1] != -1 {
			return cache[curr][last+1]
		}
		// exclude current element
		res := dfs(curr+1, last)
		// include current element if it is the first element or
		// current item is larger than last item
		if last == -1 || nums[last] < nums[curr] {
			res = max(res, 1+dfs(curr+1, curr))
		}
		cache[curr][last+1] = res
		return res
	}

	return dfs(0, -1)
}
